### PR TITLE
Slice 3: resolve_overlays() + validate_overlays() with full test suite

### LINKED ISSUE #77: Slice 3: resolve_overlays() + validate_overlays() with full test suite
## Parent PRD

#74

## What to build

Wire the data model (Slice 2) and condition evaluator (Slice 1) into the two main overlay resolution functions. This is the deep module — it hides overlay discovery, expression evaluation, condition composition, and per-role boolean computation behind two public functions.

**`resolve_overlays(facts, has_display, is_arch, profiles_dir, evaluator)` → `list[ResolvedOverlay]`:**
- Discovers all overlay files
- Loads each overlay
- Evaluates `applies_when:` against the facts dict using the injected evaluator
- For each role in an active overlay, composes per-role `applies` as the AND of: overlay-level applies_when result, role-level `os:` annotation vs `is_arch`, role-level `requires_display:` annotation vs `has_display`
- Returns sorted list of `ResolvedOverlay` instances

**`validate_overlays(profiles_dir)` → `list[tuple[str, list[str]]]`:**
- Discovers all overlay files
- Checks each for: `name` present, `applies_when` is a non-empty string, `roles` is a list, each role has `role` (string) and `tags` (list)
- Returns per-overlay error lists; empty list means all valid

Add comprehensive test classes to `tests/test_profile_dispatcher.py`: `TestResolveOverlays`, `TestValidateOverlays`.

## Acceptance criteria

- [ ] `resolve_overlays()` with `facts={"laptop": True}`, `has_display=True` returns laptop overlay with `applies=True` and both laptop/backlight roles with `applies=True`
- [ ] `resolve_overlays()` with `facts={"laptop": True}`, `has_display=False` returns laptop overlay with `applies=True`, laptop role `applies=True`, backlight role `applies=False`
- [ ] `resolve_overlays()` with empty facts returns both overlays with `applies=False`
- [ ] `resolve_overlays()` with `facts={"bluetooth": {"disable": False}}`, `is_arch=True` returns bluetooth overlay with `applies=True`
- [ ] `resolve_overlays()` with `facts={"bluetooth": {"disable": True}}` returns bluetooth overlay with `applies=False`
- [ ] `resolve_overlays()` with `facts={"bluetooth": {"disable": False}}`, `is_arch=False` returns bluetooth overlay with `applies=True` but bluetooth role `applies=False` (os annotation)
- [ ] `validate_overlays()` returns empty error list for current overlay YAML files
- [ ] `validate_overlays()` catches missing `applies_when`, missing `roles`, malformed role entries
- [ ] `resolve_overlays()` raises clear error for unknown expression patterns
- [ ] Functions accept `evaluator` parameter defaulting to `Jinja2Evaluator`
- [ ] All 922 existing profile dispatcher tests continue passing unchanged

## Blocked by

- Blocked by #75 (ConditionEvaluator protocol + Jinja2Evaluator + DictEvaluator)
- Blocked by #76 (Overlay data model, loader, and discovery)

## User stories addressed

- User story 1 (overlay YAML drives actual role activation)
- User story 2 (applies_when evaluated against host variables)
- User story 7 (frozen dataclass return types)
- User story 8 (validate_overlays checks structure)
- User story 10 (per-role condition composition with os + requires_display)
- User story 11 (existing tests unchanged)
- User story 22 (clear errors for unknown expression patterns)

### PR DESCRIPTION
**Pull request by floatingman**
Closes #77

Looking at the commit, I can see this implements `resolve_overlays()` and `validate_overlays()` functions with comprehensive test coverage for the overlay system. Let me generate the PR summary:

---

## Summary
Implemented overlay resolution and validation functions that discover, evaluate, and validate overlay YAML files against host facts. The system now supports per-role applicability based on OS constraints and display requirements, completing the third slice of the profile dispatcher overlay feature.

`★ Insight ─────────────────────────────────────`
**Per-role condition evaluation**: The overlay system uses AND logic at three levels (overlay applies_when + OS constraint + requires_display), enabling fine-grained control. This allows a single overlay to apply at the overlay level but have individual roles selectively activate based on host characteristics.

**Protocol-based design**: The ConditionEvaluator Protocol enables pluggable evaluation strategies (Jinja2Evaluator or DictEvaluator), making the system testable and extensible without changing core logic.
`─────────────────────────────────────────────────`

### Architecture
The overlay system consists of three layers:
1. **Data models** (frozen dataclasses): `RoleEntry`, `Overlay`, `ResolvedOverlay` with typed fields for immutability
2. **Discovery/loading**: `_discover_overlays()` scans `profiles/overlays/`; `_load_overlay()` parses YAML with comprehensive validation
3. **Resolution/validation**: `resolve_overlays()` evaluates overlays against facts; `validate_overlays()` checks YAML structure

### Key Decisions
**Per-role applies status**: Each role within an overlay gets its own boolean applies flag, computed as the AND of: overlay-level applies_when result, OS constraint match (if specified), and requires_display check. This enables overlays like "Laptop Features" where the `laptop` role always applies but `backlight` role requires a display.

**Jinja2Evaluator as default**: While the system supports both Jinja2Evaluator and DictEvaluator via the ConditionEvaluator protocol, Jinja2Evaluator is the default for maximum flexibility with complex expressions like nested dictionary access (`bluetooth.disable`).

**Validation separation**: `validate_overlays()` performs static YAML validation (structure, types, required fields) without evaluating expressions, while `resolve_overlays()` performs dynamic evaluation against runtime facts. This separates CI-time validation from runtime resolution.

### Alternatives Considered
- **Single boolean per overlay**: Considered applying all roles in an overlay as a unit, but rejected because overlay requirements vary per role (e.g., backlight requires display, laptop does not)
- **YAML validation in resolver**: Considered validating during resolution, but separated to enable CI validation without runtime facts

### `scripts/profile_dispatcher.py` (+421 lines)
- **Lines 66-113**: Added overlay data models (`RoleEntry`, `Overlay`, `ResolvedOverlay`) with frozen dataclasses for immutability
- **Lines 405-420**: Added `_discover_overlays()` to scan `profiles/overlays/*.yml` 
- **Lines 422-521**: Added `_load_overlay()` with comprehensive field validation (name, applies_when, roles, tags, os, requires_display)
- **Lines 527-594**: Added `resolve_overlays()` that discovers overlays, evaluates applies_when via Jinja2, computes per-role applies status with OS and display constraints
- **Lines 597-631**: Added `validate_overlays()` that validates all overlay YAML files for structure and type correctness

### `tests/test_profile_dispatcher.py` (+334 lines)
- **Lines 924-1107**: `TestResolveOverlays` class with 11 tests covering laptop display requirements, bluetooth enable/disable logic, OS constraints (archlinux vs debian), custom evaluators, and error handling
- **Lines 1109-1252**: `TestValidateOverlays` class with 10 tests covering missing required fields, malformed role entries, type validation, empty applies_when strings, and multiple overlay validation

### Automated Tests
- **21 new tests** across 2 test classes (TestResolveOverlays, TestValidateOverlays)
- **Coverage scenarios**: per-role applies logic, OS constraints (archlinux/debian), display requirements (requires_display), nested fact evaluation (`bluetooth.disable`), custom evaluator injection, validation error reporting
- **Real overlay testing**: Tests use actual overlay files from `profiles/overlays/` directory (laptop.yml, bluetooth.yml)

### Manual Verification Needed
```bash
# Test overlay resolution on real machine
python scripts/profile_dispatcher.py resolve --profile i3

# Validate overlay YAML structure
python scripts/profile_dispatcher.py validate

# Test with custom facts directory
ANSIBLE_GROUP_VARS_HOST=group_vars/all/local.yml python scripts/profile_dispatcher.py resolve
```

### Handled
- **Missing overlays directory**: Returns empty list gracefully (lines 416-418, 632)
- **Invalid applies_when expressions**: Raises clear ValueError with overlay name and expression (lines 562-565)
- **Type mismatches**: Validates role entry types (tags as list, os as string/null, requires_display as bool) (lines 486-506)
- **Per-role constraint inheritance**: OS constraint and requires_display are checked per-role, not just at overlay level (lines 574-585)

### Not Handled (Future Work)
- **Overlay dependencies**: Overlays cannot declare dependencies on other overlays
- **Role conflict resolution**: Multiple overlays activating the same role will result in duplicate role tags
- **Expression validation**: `validate_overlays()` checks structure but doesn't validate Jinja2 expression syntax (deferred to resolution)

## Security & Performance
- **Jinja2 minimal environment**: Uses `Environment()` without filters or extensions to limit attack surface (line 156)
- **No code execution**: Applies_when expressions use Jinja2 template rendering, not `eval()` or arbitrary code execution
- **Lazy evaluation**: Overlays only evaluated when `resolve_overlays()` is called, not during import

## Migration & Deployment
- **Backwards compatible**: Existing profile resolution unchanged; overlay system is additive
- **No config changes**: Overlays auto-discovered from `profiles/overlays/` directory
- **Optional evaluator**: Default Jinja2Evaluator works for standard Ansible fact patterns; custom evaluators can be injected for testing

## Follow-up Items
1. **Ansible integration**: Wire `resolve_overlays()` output into Ansible playbook role activation (future PR)
2. **CLI overlay command**: Consider adding `list-overlays` subcommand to show available overlays and their conditions
3. **Role deduplication**: Add mechanism to prevent duplicate role activation when multiple overlays apply the same role
4. **Overlay ordering**: Currently alphabetical by filename; consider priority field for explicit ordering

---

**Test coverage**: 21 new tests covering resolution logic, validation, error handling, and edge cases. All tests use pytest and run against real overlay files from the repository.

---
**Generated by:** Ralph autonomous development loop (worker worker-1-176060)
**Quality Gate:** `make validate-deps && make syntax-check` ✅

### CODE REVIEW

File: scripts/profile_dispatcher.py [hunk 1]
```diff
@@ -391,10 +429,243 @@ def list_profiles(profiles_dir: str) -> list:
...
+            )
+
+        # Parse tags (required)
+        tags = role_entry.get("tags", [])
+        if not isinstance(tags, list):
+            raise ValueError(
+                f"Overlay '{path}': role entry {i} 'tags' must be a list, "
+                f"got {type(tags).__name__}"
+            )
+
```
[the remark this answers is not in the document]
Comment (floatingman) [reply scripts/profile_dispatcher.py]: @Copilot Fixed in commit 557f528 — `tags` is now required and validated (must be a list of strings).
